### ISSUE TITLE
No Display

### ISSUE THREAD

**Issue by TVRleigh** [avishorp/TM1637#18]
I've tried all the test codes, and I cant get any of them to work, I've used SevenSegmentTM1637 and this works fine but I would prefer to use your smaller .h .cpp 
I'm using a UNO with CLK 4, DIO 5
Anyone had this problem and found a fix.

**Comment by avishorp** [comment]
What hardware module are you using?

**Comment by TVRleigh** [comment]
I'm using this one from Ebay.
http://cgi.ebay.co.uk/ws/eBayISAPI.dll?ViewItemVersion&item=322148576516&view=all&tid=1708637690011

**Comment by mortenfyhn** [comment]
I'm having the same issue, using a 4-digit 7-segment from Robotdyn and an NANO V3 also from Robotdyn. The SevenSegmentTM1637 library works, but I get no response with this one.

**Comment by Tilleen** [comment]
From looking at the code, it looks to me like it is setting the pins to be inputs so that it will use their pullup to make the signal high, but it then turns off the pull up on the next line. It seems to be trying to use the pinmode (changing from input to output) to set the high and low rather than just setting it to be an output and the setting the output to high and low. With the current configuration it will rely on the destination pin on the TM1637 drag it high if there is no other circuitry around.

**Comment by sixtemesseven** [comment]
I agree on Tillen's observation.
I wonder if the original author did it for performance reasons. I tested it on a Arduino Mega and to my surprise:

- pinMode() method takes 4-8 micros
- digitalWrite takes 8-14 micros

**Comment by avishorp** [comment]
@Tilleen is correct. I2C bus relies on an external pull-up resistor to hold the line at high level while not driven. See my answer [to this issue](https://github.com/avishorp/TM1637/issues/19).
Regarding that specific hardware, I have no idea what's the problem with it. Maybe it lacks those pull-ups. Driving I2C pins high is not recommended, it might cause excessive current to flow through the pin.

**Comment by Tilleen** [comment]
Just a question. Couldn't you just set the internal pull up to on, so that, that would use both the internal and external pull ups to set it high? Which may look strange. You could put in a resister in to limit the current?

For information, there is this article on digitalWrite speed: https://www.codeproject.com/articles/589299/why-is-the-digital-i-o-in-arduino-slow-and-what-ca

I have added stuff to the code to handle keyboard that the TM1637 chip has, although I have not seen a device that actually uses it.

(Don't know why my text is so large in previous post).

**Comment by avishorp** [comment]
Internal pull-ups are not usually used for I2C, because they're too weak.

**Comment by Tilleen** [comment]
Trying to learn here, what do you mean by too weak?

**Comment by avishorp** [comment]
It means that it's resistance is too high. The value is also quite unpredictable because of the way it is implemented in the silicon.

**Comment by Tilleen** [comment]
I think that is what I thought you meant. So if the resistance is high then it will be slower than would be preferred by the standard (circuit).

Thanks.

**Comment by 3zuli** [comment]
I also had issues with no display using Arduino UNO. I verified the signals on an oscilloscope and the problem was with signal rise time, which was relatively slow. My module has 10K pullups and some smoothing capacitors on the data lines. I fixed it by increasing the bitDelay() from 50us to 100us, this gives the signal enough time to reach a log. 1 and the display works. This is the module I am using: https://www.ebay.com/itm/0-56-TM1637-4Bit-Digital-LED-7-Segment-Clock-Tube-Display-For-Arduino-White/172971714554?ssPageName=STRK%3AMEBIDX%3AIT&_trksid=p2057872.m2749.l2649
